File: Leetcode Daily Challenge/027-September-2025/023-compare-version-numbers.cpp
// https://leetcode.com/problems/compare-version-numbers/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    vector<int> getRevision(string &s) {
        vector<int> rev;
        istringstream ss(s);
        string str;
        while (getline(ss, str, '.')) {
            rev.push_back(stoi(str));
        }
        return rev;
    }

    // TC : O(M + N)    // Where M = v1.length() && N = v2.length()
    // SC : O(M + N)
    int way_1(string &v1, string &v2) {
        vector<int> rev1 = getRevision(v1);
        vector<int> rev2 = getRevision(v2);

        int i = 0, j = 0;
        while (i < rev1.size() || j < rev2.size()) {
            int val1 = (i < rev1.size() ? rev1[i] : 0);
            int val2 = (j < rev2.size() ? rev2[j] : 0);
            if (val1 < val2) {
                return -1;
            } else if (val1 > val2) {
                return 1;
            }
            i ++, j ++;
        }
        return 0;
    }

    // TC : O(M + N)
    // SC : O(1)
    int way_2(string &v1, string &v2) {
        int m = v1.length(), n = v2.length();

        int val1 = 0, val2 = 0;
        for (int i = 0, j = 0; i < m || j < n; i ++, j ++) {
            while (i < m && v1[i] != '.') {
                val1 = val1 * 10 + (v1[i] - '0');
                i ++;
            }
            while (j < n && v2[j] != '.') {
                val2 = val2 * 10 + (v2[j] - '0');
                j ++;
            }
            if (val1 < val2) return -1;
            else if (val1 > val2) return 1;
            val1 = val2 = 0;
        }
        return 0;
    }
public:
    int compareVersion(string version1, string version2) {
        // return way_1(version1, version2);
        return way_2(version1, version2);
    }
};

File: Leetcode Daily Challenge/027-September-2025/023-compare-version-numbers.java
// https://leetcode.com/problems/compare-version-numbers/




import java.util.*;

class Solution {
    public int compareVersion(String version1, String version2) {
        // return way_1(version1, version2);
        return way_2(version1, version2);
    }

    // TC : O(M + N)    // M = rev1.size(), N = rev2.size()
    // SC : O(1)    
    private static int way_2(String v1, String v2) {
        int m = v1.length(), n = v2.length();
        int val1 = 0, val2 = 0;

        for (int i = 0, j = 0; i < m || j < n; i ++, j ++) {
            while (i < m && v1.charAt(i) != '.') {
                val1 = val1 * 10 + v1.charAt(i) - '0';
                i ++;
            }
            while (j < n && v2.charAt(j) != '.') {
                val2 = val2 * 10 + v2.charAt(j) - '0';
                j ++;
            }
            if (val1 < val2) return -1;
            else if (val1 > val2) return 1;
            val1 = val2 = 0;
        }

        return 0; 
    }

    // TC : O(M + N)    // M = rev1.size(), N = rev2.size()
    // SC : O(M + N)    
    private static int way_1(String v1, String v2) {
        List<Integer> rev1 = getRevision(v1);
        List<Integer> rev2 = getRevision(v2);

        int m = rev1.size(), n = rev2.size();
        int val1 = 0, val2 = 0;

        for (int i = 0, j = 0; i < m || j < n; i ++, j ++) {
            val1 = (i < m ? rev1.get(i) : 0);
            val2 = (j < n ? rev2.get(j) : 0);

            if (val1 < val2) return -1;
            else if (val1 > val2) return 1;
        }

        return 0;
    }

    private static List<Integer> getRevision(String s) {
        List<Integer> rev = new ArrayList<>();
        String parts[] = s.split("\\.");
        for (String part : parts) {
            rev.add(Integer.parseInt(part));
        }
        return rev;
    }
}